### PR TITLE
Move ShardTensor ring ops to functional collectives; compile-safe ring attention and ball query

### PR DESCRIPTION
**Pull request by coreyjadams**
# PhysicsNeMo Pull Request

Moves the ring collectives used by ShardTensor
attention, kNN and ball query onto functional collectives, and makes the
sharded attention and radius search usable under `torch.compile` and bf16.

- `ring.py`: `perform_ring_iteration_funcol` / `finish_ring_iteration`
  (all-to-all with two nonzero splits; overlap by issuing before compute and
  waiting after). The stream-based `perform_ring_iteration_async` and
  `get_comm_stream` are removed.
- Ring SDPA forward/backward now use `funcol` collectives, without comm streams,
  events or double buffers. `ring_sdpa` runs as an eager graph-break region
  under `torch.compile`.  This can probably be addressed but hasn't been 
  tested yet.
- Sharded SDPA in bf16/fp16 (ring and replicated-query): inputs are cast to
  the active autocast dtype before the autograd Function, and the output
  saved for backward is returned in the input dtype and in the attention
  kernel's memory layout. Care is taken to use the right memory format
  and precision for these kenrels, too, or it leads to NaN in the backward
  pass.
- Sharded `radius_search` traces end to end under `torch.compile`: the ring
  ball query's per-block merge was a raw warp launch and is now an opaque
  custom op with a fake impl.
- `torch.unbind` on a `Partial` ShardTensor resolves the pending reduction
  first; `aten.detach_` is handled.
- Tests: ring iteration parametrized over the three transports; SDPA tests
  run in fp32 and bf16 autocast with O(1)-gradient losses; sharded ball-query
  compile test; unbind and detach tests.
## Checklist

- [X] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/physicsnemo/blob/main/CONTRIBUTING.md).
- [X] New or existing tests cover these changes.
- [X] The documentation is up to date with these changes.
- [X] The [CHANGELOG.md](https://github.com/NVIDIA/physicsnemo/blob/main/CHANGELOG.md) is up to date with these changes.
- [ ] An [issue](https://github.com/NVIDIA/physicsnemo/issues) is linked to this pull request.
- [ ] If I am implementing a new model or modifying any existing model, I have followed the [Models Implementation Coding Standards](https://github.com/NVIDIA/physicsnemo/wiki/Coding-standards-for-models-implementation).

## Review Process

**All PRs are reviewed by the PhysicsNeMo team before merging.**

Depending on which files are changed, GitHub may automatically assign a maintainer for review.

We are also testing AI-based code review tools (e.g., Greptile), which may add automated comments with a confidence score.
This score reflects the AI’s assessment of merge readiness and is **not** a qualitative judgment of your work, nor is
it an indication that the PR will be accepted / rejected.

AI-generated feedback should be reviewed critically for usefulness.
You are not required to respond to every AI comment, but they are intended to help both authors and reviewers.
Please react to Greptile comments with 👍 or 👎 to provide feedback on their accuracy.

---

<sub>Stack created with <a href="https://github.com/github/gh-stack">GitHub Stacks CLI</a> • <a href="https://gh.io/stacks-feedback">Give Feedback 💬</a></sub>

### CODE REVIEW

File: physicsnemo/domain_parallel/shard_utils/ring.py [hunk 1]
```diff
@@ -24,20 +24,24 @@
...
 
 from __future__ import annotations
 
+import math
 from dataclasses import dataclass
 from typing import Literal
 
 import torch
 import torch.distributed as dist
+import torch.distributed._functional_collectives as funcol
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/ring.py:44]: This file change is really not a change in _algorithm_ at all.  it's a move from cuda.streams to funcol collectives to `ring.py` so that we can compile it.  Instead of explicit overlap, we use a launch-and-wait paradigm.

Note that totally disables the P2P route which is called out in the docstring.

File: physicsnemo/domain_parallel/shard_utils/ring.py [hunk 2]
```diff
@@ -93,42 +97,6 @@ def __post_init__(self) -> None:
             )
 
 
-_comm_streams: dict[int, torch.cuda.Stream] = {}
-
-
-def get_comm_stream(device: torch.device | int) -> torch.cuda.Stream:
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/ring.py:99]: No need to get streams anymore

File: physicsnemo/domain_parallel/shard_utils/ring.py [hunk 3]
```diff
@@ -269,89 +237,82 @@ def perform_ring_iteration(
     return tensor_recv
 
 
-def perform_ring_iteration_async(
+def perform_ring_iteration_funcol(
     tensor: torch.Tensor,
     mesh: DeviceMesh,
     ring_config: RingPassingConfig,
-    recv_tensor: torch.Tensor | None = None,
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/ring.py:276]: THere is no need to send in the receive buffer explicitly when using funcol, we can allocate it internally here.

File: physicsnemo/domain_parallel/shard_utils/ring.py [hunk 4]
```diff
@@ -269,89 +237,82 @@ def perform_ring_iteration(
...
     recv_shape: torch.Size | None = None,
-) -> tuple[torch.Tensor, list[dist.Work]]:
-    r"""Non-blocking single step of ring collective communication.
+    wait: bool = True,
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/ring.py:245]: Passing wait=True makes this a synchronous function

File: physicsnemo/domain_parallel/shard_utils/ring.py [hunk 5]
```diff
@@ -269,89 +237,82 @@ def perform_ring_iteration(
...
+    flat_recv: torch.Tensor,
+    recv_shape: torch.Size,
+) -> torch.Tensor:
+    r"""Complete an in-flight ring step started with ``wait=False``.
+
+    Waits on the functional collective and reshapes the flat buffer to
+    ``recv_shape``. This is the synchronization point: place it after the
+    compute that should overlap the communication.
+    """
+    return funcol.wait_tensor(flat_recv).reshape(recv_shape)
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/ring.py:318]: This helper function is needed to finish the ring later, if we are using the async version.

File: physicsnemo/domain_parallel/shard_utils/point_cloud_ops.py
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/point_cloud_ops.py:499]: ball query here is a synchronous function still, FYI.

File: physicsnemo/domain_parallel/shard_utils/knn.py
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/knn.py:136]: knn is a synchronous function still (wait=True by default)

File: physicsnemo/domain_parallel/shard_utils/attention_patches.py [hunk 6]
```diff
@@ -36,14 +36,65 @@
...
-    perform_ring_iteration,
-    perform_ring_iteration_async,
+    finish_ring_iteration,
+    perform_ring_iteration_funcol,
 )
 
 aten = torch.ops.aten
 
 
+def _apply_autocast(*tensors: torch.Tensor) -> tuple[torch.Tensor, ...]:
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/attention_patches.py:46]: This autocasting is necessary especially for FLARE.  We end up with NaNs if we do not do this: we get funky precision errors without this.

File: physicsnemo/domain_parallel/shard_utils/attention_patches.py [hunk 7]
```diff
@@ -36,14 +36,65 @@
...
+        The inputs, cast where they are floating point and autocast is active.
+    """
+    device_type = tensors[0].device.type
+    if not torch.is_autocast_enabled(device_type):
+        return tensors
+    dtype = torch.get_autocast_dtype(device_type)
+    return tuple(t.to(dtype) if t.is_floating_point() else t for t in tensors)
+
+
+def _kernel_output_layout(t: torch.Tensor) -> torch.Tensor:
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/attention_patches.py:73]: This one was nasty to debug.  sdpa has expectations for layout that we have to preserve,  So we wrap it here to maintain it the right way the memory layout for the attention kernel.  Yuck.

File: physicsnemo/domain_parallel/shard_utils/attention_patches.py [hunk 8]
```diff
@@ -199,51 +251,23 @@ def forward(
...
-                    # Prevent the allocator from recycling current_k/v while
-                    # comm_stream is still reading them for the send.
-                    current_k.record_stream(comm_stream)
-                    current_v.record_stream(comm_stream)
+                    next_k_flat = perform_ring_iteration_funcol(
+                        current_k, mesh, ring_config, wait=False
+                    )
+                    next_v_flat = perform_ring_iteration_funcol(
+                        current_v, mesh, ring_config, wait=False
+                    )
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/attention_patches.py:270]: Note here that the ring collectives are *overlapped*

File: physicsnemo/domain_parallel/shard_utils/attention_patches.py [hunk 9]
```diff
@@ -279,20 +303,17 @@ def forward(
...
-
-                # Record completion on comm_stream and make compute_stream
-                # wait so the next iteration reads valid recv buffer data.
-                comm_stream.record_event(comm_done)
-                compute_stream.wait_event(comm_done)
-
-                current_k = next_k_buf
-                current_v = next_v_buf
+                current_k = finish_ring_iteration(next_k_flat, k.shape)
+                current_v = finish_ring_iteration(next_v_flat, v.shape)
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/attention_patches.py:307]: And here we explicitly wait on the next iteration's collectives to finish (but not on the last iteration, don't need it there).

File: physicsnemo/domain_parallel/shard_utils/attention_patches.py [hunk 10]
```diff
@@ -475,33 +462,19 @@ def backward(
...
+                next_grad_v_flat = perform_ring_iteration_funcol(
+                    grad_v, mesh, ring_config, wait=False
+                )
             else:
-                # Last iteration: blocking shift to place grads at the right rank
-                grad_k = perform_ring_iteration(grad_k, mesh, ring_config)
-                grad_v = perform_ring_iteration(grad_v, mesh, ring_config)
+                # Last iteration: shift grads to the owning rank; nothing left
+                # to overlap with, so wait immediately.
+                grad_k = perform_ring_iteration_funcol(grad_k, mesh, ring_config)
```
Comment (coreyjadams) [inline physicsnemo/domain_parallel/shard_utils/attention_patches.py:476]: Like the comment says: the last iteration in the ring does not launch async. it blocks immediately.
